### PR TITLE
Add product count visibility option to mini cart block

### LINKED ISSUE #48528: [Enhancement]: Mini cart "product count" missing in editor
### Describe the solution you'd like

When using the site editor to change the Mini-cart block's styles, one of the options is the color of the "Product count".

In the editor, the product count does not display, so the merchant can't see how the changes look unless they go to the front-end and view the site. It should be added so the changes can be previewed in real time.

image

### Describe alternatives you've considered

_No response_

### Additional context

It might be nice to consider allowing the merchant to toggle the product count visibility for the block itself:
- Always show (even with 0)
- Never show
- Show only when > 0

### PR DESCRIPTION
**Pull request by amitraj2203**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- Added functionality to display the product count badge in the site editor, enabling merchants to see style changes in real time without needing to navigate to the front-end.
- Provided merchants with options to toggle the product count visibility directly within the block settings:
  - Always show (even when the count is 0)
  - Never show
  - Show only when the count is greater than 0
- In the editor:
  - If `Always show (even when the count is 0)` is selected, it displays 0 as the product count badge.
  - If `Never show` is selected, the product count badge will not be displayed.
  - If `Show only when the count is greater than 0` is selected, it displays 2 as the product count badge.
- Note: The editor does not display the real count value, but a placeholder to indicate how it will look on the front-end.

Closes #48528 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add the `Mini Cart` block and select it.
2. In the Inspector Control, locate the Select Control for `PRODUCT COUNT`.
3. There will be three options available. Selecting any one of these will update the block in **both the editor and the front-end** accordingly, displaying the badge on the mini cart icon:
4. For example, if we select `Always show (even when the count is 0)`, it will display the product count badge even if the count is 0, in both the editor and the front-end.
5. If we select `Never show`, the product count badge will not be displayed, regardless of the count.
6. If we select `Show only when the count is greater than 0`, the badge will be displayed only if the count is greater than 0.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/index.tsx [hunk 1]
```diff
@@ -36,6 +38,21 @@ const QuantityBadge = ( {
 		}
 	}
 
+	function determineCount( count: number, visibility: string ) {
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/index.tsx]: This feels quite complex, maybe we could do something like this:

`const shouldDisplayCount = productCountVisibility === 'always' || ( productCountVisibility === 'greater_than_zero' && count > 0 );`

instead?

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx [hunk 2]
```diff
@@ -76,18 +84,34 @@ const Edit = ( { attributes, setAttributes }: Props ): ReactElement => {
 		},
 	};
 
+	const { cartItemsCount: cartItemsCountFromApi, cartIsLoading } =
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: I don't feel we should be using `useStoreCart` in the editor, maybe we could just show a number depending on what option the merchant has selected?

`never`: show nothing
`always`: 0
`greater_than_zero`:  2

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx [hunk 3]
```diff
@@ -229,10 +285,11 @@ const Edit = ( { attributes, setAttributes }: Props ): ReactElement => {
 						</span>
 					) }
 					<QuantityBadge
-						count={ productCount }
+						count={ cartItemsCount }
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: instead of adding these props to the `QuantityBadge` could we determine whether the badge should be displayed at this level and conditionally render it?

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss [hunk 4]
```diff
@@ -25,8 +25,8 @@
 
 :where(.wc-block-mini-cart__badge) {
 	// These values will be overridden in JS.
-	background-color: transparent;
-	color: transparent;
+	background-color: #111;
+	color: #f9f9f9;
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss]: What is the purpose of these changes?

**Review by opr — CHANGES_REQUESTED** [review]
Hey @amitraj2203 thanks for your contribution. This is a great start.

I recommend updating the "Changes proposed in this Pull Request" section to more thoroughly describe what you're adding, and also add some more detailed steps to the testing instructions. What you have there is good but we could be more explicit, e.g.

- Select 'Always show count (even if 0)'
- Ensure the product count shows up on the mini cart icon"

I also added some inline comments on the code. To summarise what I think we should do:

- Render `QuantityBadge` conditionally (rather than doing the check _inside_ this component)
- Show the quantity badge in the editor depending on setting's value (don't use `useStoreCart`)

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss (hunk 4 above)
Comment (amitraj2203) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss]: With the transparent value, the color of the quantity badge is not visible. I have attached a video for an explanation.

https://github.com/woocommerce/woocommerce/assets/77401999/ddc1d05f-95b3-475b-b53c-c25acd2599e0

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx (hunk 3 above)
Comment (amitraj2203) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: Since the `QuantityBadge` component includes both the cart icon and the product count badge.
So we can create a boolean prop like `showBadge` in the parent component and pass it down to it to determine whether the span with the class `wc-block-mini-cart__badge` should be visible. I'm not sure if this is the right approach. If you have any other suggestions, please let me know.

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx (hunk 2 above)
Comment (amitraj2203) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: Thanks for the suggestion, I have updated it accordingly.

**Comment by amitraj2203** [comment]
Hi @opr, thanks for reviewing the PR. I have updated the PR description and also the feedback on the code.

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx [hunk 5]
```diff
@@ -87,7 +90,20 @@ const Edit = ( { attributes, setAttributes }: Props ): ReactElement => {
 		''
 	) as string;
 
-	const productCount = 0;
+	const getDisplayCount = () => {
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: I think this function adds complexity, the value could be derived like so:

`const productCount = productCountVisibility === 'never' || productCountVisibility === 'always' ? 0 : 2;`

I concede ternary operators are a little harder to read, but in this case I feel the conditions are simple enough to make it OK.

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss (hunk 4 above)
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss]: Thanks, I see the issue now!

I wonder if we could undo this change, and add something like this in the `edit.tsx` file

 (similar logic is used in the mini-cart contents edit.tsx file)

```js
/**
 * This is a workaround for the Site Editor to set the correct
 * background color of the Mini-Cart QuantityBadge block based on
 * the main background color set by the theme.
 */
useEffect( () => {
	const canvasEl = document.querySelector(
		'.edit-site-visual-editor__editor-canvas'
	);
	if ( ! ( canvasEl instanceof HTMLIFrameElement ) ) {
		return;
	}
	const canvas =
		canvasEl.contentDocument || canvasEl.contentWindow?.document;
	if ( ! canvas ) {
		return;
	}
	if (
		canvas.getElementById( 'mini-cart-quantity-badge-foreground-color' )
	) {
		return;
	}
	const backgroundColor = window.getComputedStyle(
		canvas.body
	)?.backgroundColor;
	const color = window.getComputedStyle( canvas.body )?.color;

	if ( ! backgroundColor || ! color ) {
		return;
	}
	const style = document.createElement( 'style' );
	style.id = 'mini-cart-quantity-badge-foreground-color';

	// The badge's background should be the theme's foreground color, and it's text color the theme's background color.
	style.appendChild(
		document.createTextNode(
			`:where(.wc-block-mini-cart__badge) {
			color: ${ backgroundColor };
			background-color: ${ color };
		}`
		)
	);
	const body = canvas.querySelector( '.editor-styles-wrapper' );
	if ( ! body ) {
		return;
	}
	body.appendChild( style );
}, [] );
```

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/index.tsx [hunk 6]
```diff
@@ -48,7 +56,7 @@ const QuantityBadge = ( {
 				className="wc-block-mini-cart__badge"
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/index.tsx]: Can't we just skip showing this span if it shouldn't display? That's what I was getting at in my previous comments about `QuantityBadge` but I didn't realise this component also showed the icon

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx [hunk 7]
```diff
@@ -135,6 +151,42 @@ const Edit = ( { attributes, setAttributes }: Props ): ReactElement => {
 							}
 						/>
 					</BaseControl>
+					<BaseControl
+						id="wc-block-mini-cart__product-count-toggle"
+						label={ __( 'Product Count', 'woocommerce' ) }
+					>
+						<SelectControl
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: I think it might be nicer to display these options with radio inputs, that way the entire text string for the longer options can be displayed, while in a select it may be cut off. Do you agree?

**Review by opr — CHANGES_REQUESTED** [review]
Thanks a lot @amitraj2203! I really appreciate you addressing my comments 👍🏼 I think this PR is shaping up nicely.

I've left a few more comments that I think we should address. Please let me know if you need help with any of them.

Thanks!

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx (hunk 7 above)
Comment (amitraj2203) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: Yes I agree. I also tried it and here's how it will look with a few CSS (gap between the options)
image

Should I change it?

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss (hunk 4 above)
Comment (amitraj2203) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss]: Thanks for the wonderful suggestion 🙇. I have added the effect similar like this and also tested it by changing the theme bg color and it's working fine.

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx (hunk 7 above)
Comment (tarunvijwani) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: > Yes I agree. I also tried it and here's how it will look with a few CSS (gap between the options) image
> 
> Should I change it?

I think we should rephrase the wording to make it more user-friendly. 
Something like: 

Display Product Count

⃝ Always (even if 0)
⃝ Never
⃝ Only if cart has any product. 


@nikkivias Could you please help us with some design mocks or text that we should be using to control the product count visibility in Editor? 


The current implementation has the drop-down select option but we are considering changing it to radio buttons.

image

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx [hunk 8]
```diff
@@ -87,7 +91,58 @@ const Edit = ( { attributes, setAttributes }: Props ): ReactElement => {
 		''
 	) as string;
 
-	const productCount = 0;
+	/**
+	 * This is a workaround for the Site Editor to set the correct
+	 * background color of the Mini-Cart QuantityBadge block based on
+	 * the main background color set by the theme.
+	 */
```
Comment (tarunvijwani) [inline plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx:98]: I'm not sure about this workaround. Could you please help me with before and after images for this one along with the steps to reproduce it?

**Review by tarunvijwani — COMMENTED** [review]
When selecting the option to display product count, it doesn't update for me. Am I missing anything?  

https://github.com/woocommerce/woocommerce/assets/11503784/7bdb92d5-17f7-4f9e-a97e-be32dce40ea4

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx (hunk 8 above)
Comment (amitraj2203) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx:98]: Hi @tarunvijwani,

These [two CSS properties](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/blocks/mini-cart/quantity-badge/style.scss#L28-L29) are hiding the count badge in the Mini Cart block. So earlier I changed these values to hard-coded colors which isn't a good approach since the goal is to keep product count badge background to be the theme's foreground color, and it's text color should be the theme's background color. So for this reason I have used this workaround recommended by @/opr (which is also used in this [file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/blocks/mini-cart/mini-cart-contents/edit.tsx#L62-L113))

https://github.com/woocommerce/woocommerce/assets/77401999/ddc1d05f-95b3-475b-b53c-c25acd2599e0

But I see one problem, is that it works in the post editor but it doesn't work in the site editor since the content comes under iframe (which isn't the case in post editor) so I have to refine this `useEffect`

> Could you please help me with before and after images for this one along with the steps to reproduce it?

To test this:
1. Go the Side editor > Global Styles
2. Go to the `Colors` settings from the inspector control
3. Try changing the background color and text color and see the opposite will get applied to the product count badge to keep the styles consistent.
4. This will work in any posts where Mini Cart block is added. (For it to be work on site editor, I'll have to change in useEffect)

**Comment by amitraj2203** [comment]
> When selecting the option to display product count, it doesn't update for me. Am I missing anything?
> 
>  Screen.Recording.2024-06-20.at.1.36.59.PM.mov

I have updated the `useEffect` and it should work in the site editor too.

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx (hunk 8 above)
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx:98]: The complexity of this code could be reduced, something like this is a little easier to reason with in my opinion. Also has less nesting.

```ts
/**
 * This is a workaround for the Site Editor to set the correct
 * background color of the Mini-Cart QuantityBadge block based on
 * the main background color set by the theme.
 */
useEffect( () => {
	let editorStylesWrapper = document.querySelector(
		'.editor-styles-wrapper'
	);
	// If the editor styles wrapper is not available, look in the site editor canvas for it.
	if ( ! editorStylesWrapper ) {
		const canvasEl = document.querySelector(
			'.edit-site-visual-editor__editor-canvas'
		);

		if ( ! ( canvasEl instanceof HTMLIFrameElement ) ) {
			return;
		}
		const canvas =
			canvasEl.contentDocument || canvasEl.contentWindow?.document;
		if ( ! canvas ) {
			return;
		}
		editorStylesWrapper = canvas.querySelector(
			'.editor-styles-wrapper'
		);
	}

	if ( ! editorStylesWrapper ) {
		return;
	}

	const editorBackgroundColor =
		window.getComputedStyle( editorStylesWrapper )?.backgroundColor;
	const editorColor =
		window.getComputedStyle( editorStylesWrapper )?.color;

	if (
		editorStylesWrapper &&
		! editorStylesWrapper.querySelector(
			'#mini-cart-quantity-badge-foreground-color'
		) &&
		editorBackgroundColor &&
		editorColor
	) {
		const style = document.createElement( 'style' );
		style.id = 'mini-cart-quantity-badge-foreground-color';
		style.appendChild(
			document.createTextNode(
				`:where(.wc-block-mini-cart__badge) {
						color: ${ editorBackgroundColor };
						background-color: ${ editorColor };
					}`
			)
		);
		editorStylesWrapper.appendChild( style );
	}
}, [] );
```

File: plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx (hunk 7 above)
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/blocks/mini-cart/edit.tsx]: I think going ahead with the shortened versions makes sense here, changing the section title to "Show quantity badge" and the options to:

> ⃝ Always (even if 0)
> ⃝ Never
> ⃝ Only when cart is not empty

**Review by opr — CHANGES_REQUESTED** [review]
thanks again @amitraj2203 for your work and patience here, I made a suggestion about the`useEffect` and just want a 👍🏼 from @nikkivias on the radio buttons before approving.

**Comment by amitraj2203** [comment]
> thanks again @amitraj2203 for your work and patience here, I made a suggestion about the`useEffect` and just want a 👍🏼 from @/nikkivias on the radio buttons before approving.

Acknowledged. Here's how it's looking now.
image

**Comment by dominique-vias** [comment]
Hi team thank you for the work done here so far! I only have small edits to suggest to the wording and the order of things:

Show Cart Item Count:

- Always (even if empty)
- Only if cart has items
- Never

@opr

**Comment by amitraj2203** [comment]
Thanks for the feedback.
image

**Review by opr — APPROVED** [review]
@amitraj2203 thanks a lot for your hard work and patience on this. Thanks for sticking with it while we got design input and for making the changes so quickly after our reviews.

This is working well for me and the code is looking good. The next step is to add some unit tests for this change, but we can [address that in a follow-up](https://github.com/woocommerce/woocommerce/issues/49105) so we don't have to block this PR any further.

**Comment by opr** [comment]
@amitraj2203 could you rebase this onto `woocommerce/woocommerce/trunk` please?

**Comment by amitraj2203** [comment]
@opr I have rebased it but don't know why still some test cases are failing.

**Review by dominique-vias — APPROVED** [review]

**Comment by adimoldovan** [comment]
> @opr I have rebased it but don't know why still some test cases are failing.

@amitraj2203 The checks are looking good. The Build Live Branch and Test reports jobs are known and expected to fail on forks. The Evaluate Project Job Statuses fails because of Test reports jobs failing. I'll override that and merge this PR.

**Comment by adimoldovan** [comment]
I will merge this PR by overriding the `Evaluate Project Job Statuses` failure. This is a regression that will be address with https://github.com/woocommerce/woocommerce/pull/49141

**Comment by opr** [comment]
@amitraj2203 thanks a lot for your hard work on this and sticking with it while we got the design feedback! We really appreciate you taking the time to contribute to WooCommerce 🙌🏼 

If all goes well, your changes should be included with WooCommerce 9.2.0 in August 🎊

**Comment by amitraj2203** [comment]
@opr Thanks for your kind words and appreciation 😄 ! I'm glad to have been able to contribute, and I really appreciate your support and feedback throughout the process 🙌🏼
